timelock-impl/src/main/java/com/palantir/atlasdb/timelock/transaction/timestamp/TimestampRanges.java

```diff
@@ -0,0 +1,64 @@
+/*
+ * Copyright 2018 Palantir Technologies, Inc. All rights reserved.
+ *
+ * Licensed under the BSD-3 License (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://opensource.org/licenses/BSD-3-Clause
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.timelock.transaction.timestamp;
+
+import java.util.OptionalLong;
+
+import com.google.common.base.Preconditions;
+import com.palantir.timestamp.TimestampRange;
+
+/**
+ * Utility methods for manipulating ranges of timestamps.
+ */
+public final class TimestampRanges {
+    private TimestampRanges() {
+        // utility
+    }
+
+    /**
+     * Returns a timestamp in the provided timestamp range that has the provided residue class modulo the provided
+     * modulus.
+     *
+     * We do not make any guarantees on which timestamp is provided if multiple timestamps in the range reside
+     * in the provided residue class. For example, if the TimestampRange is from 1 to 5 inclusive and we want a
+     * timestamp with residue 1 modulo 2, this method may return any of 1, 3 and 5.
+     *
+     * If the timestamp range does not contain a timestamp matching the criteria, returns empty.
+     *
+     * @param range timestamp range to extract a single timestamp from
+     * @param residue desired residue class of the timestamp returned
+     * @param modulus modulus used to partition numbers into residue classes
+     * @return a timestamp in the given range in the relevant residue class modulo modulus
+     * @throws IllegalArgumentException if residue >= modulus
+     */
+    public static OptionalLong getTimestampMatchingModulus(TimestampRange range, int residue, int modulus) {
```

Comment:
why not just put this on the `TimestampRange`? Similar to how guava's `Range` works.

Reply:
Fair enough, will do.

Reply:
I've moved this off the `TimestampRange` and back to the static class. Reason is that `TimestampRange` is in `timestamp-api` which doesn't have Guava and it doesn't feel right to force users of the timestamp API to have a Guava dependency - and I'd prefer not to copy related `LongMath` and `Preconditions` stuff into our project.

Reply:
I'm not sure I buy this, there aren't non-atlas consumers of `timestamp-api` and at the very least we're not encouraging this. So if we assume everyone who has this has atlas-client, then they're already pulling in Guava (albeit transitively). So not really sure why it matters here. e.g. `qos-service-api` pulls in an entire remoting3 which I would argue is a lot more heavy handed than guava (and *does* pull in guava anyway). 

It does seem like the best place to put it and isn't just a mere utility, but I won't push too much.

Reply:
There _are_ non-Atlas consumers of `timestamp-api` though (check large internal project). The actual server in said non-Atlas consumer does pull in Guava on its own, but IMO that's a separate concern and I'm not sure we should force that.

Regarding QoS, I agree that Remoting3 is a lot more heavy handed than Guava, but I'd be inclined to agree with the "everyone who has this has atlas-client" statement a lot more strongly if you're a QoS user. (I agree with the logic, just not so much with the premise in this case)